### PR TITLE
Refactor dashboard to use factory and new cache provider

### PR DESCRIPTION
**Pull request by galvani**
~~⚠️ Depends on https://github.com/mautic/mautic/pull/9107 which must be merged first! ⚠️~~

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | N
| Automated tests included? | Y
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR makes the dashboard use the CacheBundle instead of the cache storage.

#### Steps to test this PR:
1. Enable some adapter for the CacheBundle 

For redis I recommend getting some GUI like *Redis Desktop Manager*

2. Pull this PR and clear cache
3. Refresh the /dashboard and verify that data is being written to the cache you have selected
4. Verify cache is being displayed

### CODE REVIEW

**Comment by galvani** [comment]
## Depends on https://github.com/mautic/mautic/pull/9107

**Comment by npracht** [comment]
Hi there!
It has been decided to not create any extra Mautic 2 minor versions (which means no more features in M2) knowing that Mautic 3 is coming very soon.

**We now want to integrate your contribution in the Mautic 3 roadmap as 3.1.0 candidate.**

## How to do?
1. Check if your feature or enhancement is still missing and relevant in Mautic 3
2. **If Yes**: rebase your PR against the 3.x branch
3. **If No**: please consider closing your PR if you are absolutely sure that the feature has made it into Mautic 3 already

**Please report results by commenting on your PR to make us administration easier.**
______
You can more information on how to do all of that on [this blog post "Getting you PR ready for Mautic 3"](https://www.mautic.org/blog/developer/getting-your-pr-ready-mautic-3).

**Comment by RCheesley** [comment]
@alanhartless @escopecz another one that needs to be picked up :)

**Comment by escopecz** [comment]
This PR depends on https://github.com/mautic/mautic/pull/8817. I can get this one into a good shape once we merge that one.

**Comment by jordan8037310** [comment]
Marking that this is now dependent on #9107

**Comment by RCheesley** [comment]
I've assigned this to the 3.2 milestone - will need to get the cache bundle merged in before we can look at merging this PR.

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by anton-vlasenko** [comment]
This PR now also includes this fix and should also be tested:

[//]: # ( Required: )
#### Description:
Widget content is cached to avoid having to calculate the results each time. The problem is that the cache is used even when the date range is changed and thus the widget still shows the old cached data which is invalid for the new date range.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. open a mautic instance.
3. look at the dashboard
4. change the date range
5. look at the dashboard again and see nothing changed - the date range remains the same
6. clear cache (can be tricky, based on which cache handler is enabled. You can try to use the php console mautic:cache:clear command)
7. see that the widgets now show correct data

#### Steps to test this PR:
1. Load up [this PR]
2. look at the dashboard
3. change the date range
4. look at the dashboard again and see that the date is changed

**Comment by anton-vlasenko** [comment]
@cla-bot check

File: app/bundles/DashboardBundle/Config/config.php [hunk 1]
```diff
@@ -77,6 +78,14 @@
                     'session',
                 ],
             ],
+            'mautic.dashboard.widget_detail.event.factory' => [
+                'class'     => \Mautic\DashboardBundle\Widget\WidgetDetailEventFactory::class,
+                'arguments' => [
+                    'translator',
+                    'mautic.cache.provider',
+                    'mautic.helper.user',
```
Comment (kuzmany) [inline app/bundles/DashboardBundle/Config/config.php]: Here could be on third place 'mautic.security' service
I tried push it but do not have permission

**Review by kuzmany — CHANGES_REQUESTED** [review]

File: app/bundles/DashboardBundle/Config/config.php (hunk 1 above)
Comment (anton-vlasenko) [reply app/bundles/DashboardBundle/Config/config.php]: @kuzmany I don't have time to fix it now. But I agree with you on that.
I'll try to find time to fix it.

**Comment by kuzmany** [comment]
@escopecz can you resolve conflicts please?

**Review by kuzmany — CHANGES_REQUESTED** [review]
From code perspective and from PHPStorm I think the cache bundle is not used


image


usesLegacyCache is still true.

@escopecz can you check it?

**Comment by escopecz** [comment]
@kuzmany you are right. This PR is missing some changes from the orignal PR in our repo. None of the commits has the same commit messages as the orignal PR does and a lot of commits on top of it. Well, 2 years old PR...

**Comment by escopecz** [comment]
I hunted down the missing changes and added them in https://github.com/mautic/mautic/pull/7785/commits/e88b7c863839fdf64c81e12b775ad770c582c8d3. I don't know what happened here.

**Comment by kuzmany** [comment]
If I import default dashboards, I see these errors:

image

**Review by kuzmany — CHANGES_REQUESTED** [review]
I see this after checkout

image

**Comment by escopecz** [comment]
I will have to get to this one some time later. It is not a priority

File: app/bundles/DashboardBundle/Config/config.php [hunk 2]
```diff
@@ -42,13 +42,22 @@
...
             ],
+            'mautic.dashboard.widget_detail.event.factory' => [
+                'class'     => Mautic\DashboardBundle\Event\WidgetDetailEventFactory::class,
+                'arguments' => [
+                    'translator',
+                    'mautic.cache.provider',
+                    'mautic.security',
+                    'mautic.helper.user',
+                ],
+            ],
```
Comment (escopecz) [inline app/bundles/DashboardBundle/Config/config.php]: This can be removed as M5 comes with autowiring
```suggestion
```

File: app/bundles/DashboardBundle/Event/WidgetDetailEvent.php [hunk 3]
```diff
@@ -149,26 +175,41 @@ public function getTemplate()
 
     /**
      * Set the widget template data.
+     *
+     * @throws \Psr\Cache\InvalidArgumentException
      */
-    public function setTemplateData(array $templateData, $skipCache = false): void
+    public function setTemplateData(array $templateData, ?bool $skipCache = false): bool
```
Comment (escopecz) [inline app/bundles/DashboardBundle/Event/WidgetDetailEvent.php]: We cannot change the method signature. Pease add the return types as method annotations. Is it necessary to change the return type? It's not used anywhere as far as I can see.

If it's necessary to change the types, then please create a new method for it and mark this one as deprecated.

File: app/bundles/DashboardBundle/Event/WidgetDetailEvent.php [hunk 4]
```diff
@@ -254,14 +307,6 @@ public function getTranslator()
         return $this->translator;
     }
 
-    /**
-     * Set security object to check the perimissions.
-     */
-    public function setSecurity(CorePermissions $security): void
```
Comment (escopecz) [inline app/bundles/DashboardBundle/Event/WidgetDetailEvent.php:260]: Please, revert this change as some plugin can call this public method. Mark it as deprecated instead.

File: app/bundles/DashboardBundle/Tests/Model/DashboardModelTest.php [hunk 5]
```diff
@@ -6,60 +6,68 @@
...
-    private \PHPUnit\Framework\MockObject\MockObject $session;
+    private MockObject $session;
 
-    private \Mautic\DashboardBundle\Model\DashboardModel $model;
+    private DashboardModel $model;
+
+    /**
+     * @var WidgetDetailEventFactory
+     */
+    private $widgetDetailEventFactory;
```
Comment (escopecz) [inline app/bundles/DashboardBundle/Tests/Model/DashboardModelTest.php]:
```suggestion
    private WidgetDetailEventFactory $widgetDetailEventFactory;
```

**Review by escopecz — CHANGES_REQUESTED** [review]

File: app/bundles/DashboardBundle/Event/WidgetDetailEvent.php (hunk 3 above)
Comment (dadarya0) [reply app/bundles/DashboardBundle/Event/WidgetDetailEvent.php]: This function was returning boolean value in return statement but now I have verified that response of this function is not used anywhere so I have reverted again return type to void.

File: app/bundles/DashboardBundle/Config/config.php (hunk 2 above)
Comment (dadarya0) [reply app/bundles/DashboardBundle/Config/config.php]: By default we are ignoring autowiring for Event directory so I had added this as service but now I have loaded This factory in autowiring service file.

Comment (escopecz) [reply app/bundles/DashboardBundle/Config/config.php]: @dadarya0 I see. In that case I'd recommend to move the `WidgetDetailEventFactory` to another folder because we shouldn't be mixing service classes with DTO classes. Can you move it to another folder? Like `app/bundles/DashboardBundle/Factory/WidgetDetailEventFactory.php`?

File: app/bundles/DashboardBundle/Event/WidgetDetailEvent.php [hunk 6]
```diff
@@ -149,26 +175,39 @@ public function getTemplate()
 
     /**
      * Set the widget template data.
+     *
+     * @throws \Psr\Cache\InvalidArgumentException
      */
-    public function setTemplateData(array $templateData, $skipCache = false): void
+    public function setTemplateData(array $templateData, ?bool $skipCache = false): void
```
Comment (escopecz) [inline app/bundles/DashboardBundle/Event/WidgetDetailEvent.php]: Please add the new param type to a method comment. Changing a method signature is considered a BC break. We'll have to add the real types for M6.

File: app/bundles/DashboardBundle/Config/config.php (hunk 2 above)
Comment (dadarya0) [reply app/bundles/DashboardBundle/Config/config.php]: Moved to new folder named Factory

**Review by escopecz — APPROVED** [review]
The code looks great and it works as before. When I change the date range to enforce to load the data from database I can see it works:
Screenshot 2024-01-31 at 13 14 15
And when I refresh the data is loaded from cache as expected:
Screenshot 2024-01-31 at 13 14 32

**Review by kuzmany — APPROVED** [review]
Code looks good 🫰
Seems working properly.

Just wondering use deprecated CacheStorageHelper. I understand why, but good chance to get rid of it on this place

**Comment by escopecz** [comment]
@kuzmany good point. We'll remove it in M6.
